app/sprinkles/account/src/Authenticate/PasswordSecurity.php

```diff
@@ -0,0 +1,185 @@
+     * Checks the `flag_password_reset_required` column for a user.
+     *
+     * @param UserInterface $currentUser
+     *
+     * @return bool True if a password reset is required, false otherwise.
+     */
+    public function checkPasswordResetRequired(UserInterface $currentUser)
+    {
+        if ($currentUser->flag_password_reset_required) {
+            return true;
+        }
+
+        return false;
+    }
+
+    /**
+     * Checks if compromised password reset feature is enabeld.
+     *
+     * @return bool True if the feature is enabled.
+     */
+    public function resetCompromisedEnabled()
+    {
+        return $this->config['site.login.enforce_reset_compromised'];
+    }
+
+    /**
+     * Sets the `flag_password_reset_required` column for a user to true.
+     *
+     * @param UserInterface $currentUser
+     */
+    public function setPasswordResetRequired(UserInterface $currentUser)
+    {
+
+        /** @var \UserFrosting\Sprinkle\Core\Util\ClassMapper $classMapper */
+        $classMapper = $this->classMapper;
+
+        // All checks passed!  log events/activities, update user, and send email
+        // Begin transaction - DB will be rolled back if an exception occurs
+        Capsule::transaction(function () use ($classMapper, $currentUser) {
+
+            // Load the user, by username
+            $user = $classMapper->getClassMapping('user')::where('user_name', $currentUser['user_name'])->first();
+            $user->flag_password_reset_required = true;
+            $user->save();
+        });
+    }
+
+    /**
+     * Check a password SHA1 hash against an array of compromised password hashes.
+     *
+     * @param string $hash  The hash of the potential password to be used.
+     * @param array  $array Array of password hashes in the format c2d18a7d49b0d4260769eb03d027066d29a:181 - or <hash>:<number of breaches.
+     *
+     * @return int The number of times a password has been exposed in data breaches.
+     */
+    private function checkHash($hash, $array)
+    {
+        foreach ($array as $index => $pwHash) {
+            $breachedItemParts = explode(':', $pwHash);
+
+            $breachedItemHash = $breachedItemParts[0];
+            $numberOfBreaches = $breachedItemParts[1];
+
+            // compare the hash suffix from array of hash suffix with password hash suffix.
+            if ($breachedItemHash == substr($hash, 5)) {
+                return $breaches = (int) trim($numberOfBreaches);
+            } else {
+                $breaches = 0;
+            }
+        }
+
+        return $breaches;
+    }
+
+    /**
+     * Queries Have I been Pwned API to generate list of hashed compromised password suffixes.
+     *
+     * @param string $hashPrefix The prefix (first 5 characters) of hashed password.
+     *
+     * @return array Array of password hashes in the format c2d18a7d49b0d4260769eb03d027066d29a:181 - or <hash>:<number of breaches.
+     */
+    private function getHashArrayFromAPI($hashPrefix)
```

Comment:
You're right. 

I'll think about the best way to handle this. We probably want to try to query HIBP at least a few times before allowing user to login and bypass the check...

Reply:
Got to be careful with retry logic. At scale, it can translate to a massive spike in requests for an API.
Possible scenarios include;
* API returns 500 error due to a fault, could be an edge case or everything, retry logic here would easily result in a larger number of requests (CloudFlare wouldn't cache these either, so it would fairly consistently hit the function itself until potentially DDoS protection gets triggered).
* Response codes indicate that retries should not be attempted exist, we would be wasting our own time retrying these.
* Connection timeouts could be attributed to failures anywhere along the connection (local misconfiguration, API hit its own response timeout and terminated).
* Deliberate rate limiting by HIBP (if a UF site were to become heavily used, we could end up hitting this)

Ultimately, it may come down to not retrying at all and logging that the check failed. HIBP passwords should be supplemental to the security such that any UF site doesn't get busted by any issues.

Reply:
I've tried to fix this in the recent commits. (Also switched over to Guzzle)